File: src/main/java/me/ayydxn/luminescence/javascript/JSValue.java
package me.ayydxn.luminescence.javascript;

import me.ayydxn.luminescence.util.UsedByNative;

public class JSValue
{
    @UsedByNative("JSValue.cpp")
    final long contextHandle;

    @UsedByNative("JSValue.cpp")
    final long valueHandle;

    JSValue(long contextHandle, long valueHandle)
    {
        this.contextHandle = contextHandle;
        this.valueHandle = valueHandle;
    }

    public JSType getType()
    {
        return NativeMethods.nJSValueGetType(contextHandle, valueHandle);
    }

    public boolean isUndefined()
    {
        return this.getType() == JSType.UNDEFINED;
    }

    public boolean isNull()
    {
        return this.getType() == JSType.NULL;
    }

    public boolean isBoolean()
    {
        return this.getType() == JSType.BOOLEAN;
    }

    public boolean isNumber()
    {
        return this.getType() == JSType.NUMBER;
    }

    public boolean isString()
    {
        return this.getType() == JSType.STRING;
    }

    public boolean isObject()
    {
        return NativeMethods.nJSValueIsObject(contextHandle, valueHandle);
    }

    public boolean isArray()
    {
        return NativeMethods.nJSValueIsArray(contextHandle, valueHandle);
    }

    public boolean isFunction()
    {
        return NativeMethods.nJSValueIsFunction(contextHandle, valueHandle);
    }

    public boolean toBoolean()
    {
        return NativeMethods.nJSValueToBoolean(contextHandle, valueHandle);
    }

    public double toNumber() throws JSException
    {
        return NativeMethods.nJSValueToNumber(contextHandle, valueHandle);
    }

    public String toJSONString(int indent) throws JSException
    {
        return NativeMethods.nJSValueToJSONString(contextHandle, valueHandle, indent);
    }

    @Override
    public String toString()
    {
        return NativeMethods.nJSValueToString(contextHandle, valueHandle);
    }

    private static final class NativeMethods
    {
        private static native JSType nJSValueGetType(long contextHandle, long valueHandle);

        private static native boolean nJSValueIsObject(long contextHandle, long valueHandle);

        private static native boolean nJSValueIsArray(long contextHandle, long valueHandle);

        private static native boolean nJSValueIsFunction(long contextHandle, long valueHandle);

        private static native boolean nJSValueToBoolean(long contextHandle, long valueHandle);

        private static native double nJSValueToNumber(long contextHandle, long valueHandle) throws JSException;

        private static native String nJSValueToJSONString(long contextHandle, long valueHandle, int indent) throws JSException;

        private static native String nJSValueToString(long contextHandle, long valueHandle);
    }
}

File: src/main/native/JavaScript/JSValue.cpp
#include "JSCUtilities.h"
#include "Cache/JSCCache.h"
#include "Core/JNIUtilities.h"

#include <JavaScriptCore/JavaScript.h>

using namespace Luminescence::JavaScript;

namespace
{
    /*
     * Class:     me_ayydxn_luminescence_javascript_JSValue_NativeMethods
     * Method:    nJSValueGetType
     * Signature: (JJ)Lme/ayydxn/luminescence/javascript/JSType;
     */
    jobject JNICALL JSValueGetType_Native(JNIEnv* Environment, jclass, jlong ContextHandle, jlong ValueHandle)
    {
        const int ValueType = JSValueGetType(ToContextRef(ContextHandle), ToValueRef(ValueHandle));
        
        return Environment->CallStaticObjectMethod(CJSTypeCache::ClassRef, CJSTypeCache::FromNativeMethodID, ValueType);
    }

    /*
     * Class:     me_ayydxn_luminescence_javascript_JSValue_NativeMethods
     * Method:    nJSValueIsObject
     * Signature: (JJ)Z
     */
    jboolean JNICALL JSValueIsObject_Native(JNIEnv*, jclass, jlong ContextHandle, jlong ValueHandle)
    {
        return JSValueIsObject(ToContextRef(ContextHandle), ToValueRef(ValueHandle));
    }

    /*
     * Class:     me_ayydxn_luminescence_javascript_JSValue_NativeMethods
     * Method:    nJSValueIsArray
     * Signature: (JJ)Z
     */
    jboolean JNICALL JSValueIsArray_Native(JNIEnv*, jclass, jlong ContextHandle, jlong ValueHandle)
    {
        return JSValueIsArray(ToContextRef(ContextHandle), ToValueRef(ValueHandle));
    }

    /*
     * Class:     me_ayydxn_luminescence_javascript_JSValue_NativeMethods
     * Method:    nJSValueIsFunction
     * Signature: (JJ)Z
     */
    jboolean JNICALL JSValueIsFunction_Native(JNIEnv*, jclass, jlong ContextHandle, jlong ValueHandle)
    {
        const auto ContextRef = ToContextRef(ContextHandle);

        return JSObjectIsFunction(ContextRef, JSValueToObject(ContextRef, ToValueRef(ValueHandle), nullptr));
    }

    /*
     * Class:     me_ayydxn_luminescence_javascript_JSValue_NativeMethods
     * Method:    nJSValueToBoolean
     * Signature: (JJ)Z
     */
    jboolean JNICALL JSValueToBoolean_Native(JNIEnv*, jclass, jlong ContextHandle, jlong ValueHandle)
    {
        return JSValueToBoolean(ToContextRef(ContextHandle), ToValueRef(ValueHandle));
    }

    /*
     * Class:     me_ayydxn_luminescence_javascript_JSValue_NativeMethods
     * Method:    nJSValueToNumber
     * Signature: (JJ)D
     */
    jdouble JNICALL JSValueToNumber_Native(JNIEnv* Environment, jclass, jlong ContextHandle, jlong ValueHandle)
    {
        JSValueRef Exception = nullptr;

        const jdouble Value = JSValueToNumber(ToContextRef(ContextHandle), ToValueRef(ValueHandle), &Exception);
        if (ThrowIfJSException(Environment, ToContextRef(ContextHandle), Exception))
            return 0.0;

        return Value;
    }

    /*
     * Class:     me_ayydxn_luminescence_javascript_JSValue_NativeMethods
     * Method:    nJSValueToJSONString
     * Signature: (JJI)Ljava/lang/String;
     */
    jstring JNICALL JSValueToJSONString_Native(JNIEnv* Environment, jclass, jlong ContextHandle, jlong ValueHandle,
                                               jint Indent)
    {
        JSValueRef Exception = nullptr;

        const JSStringRef JsonString = JSValueCreateJSONString(ToContextRef(ContextHandle), ToValueRef(ValueHandle),
                                                               static_cast<unsigned>(Indent), &Exception);
        if (ThrowIfJSException(Environment, ToContextRef(ContextHandle), Exception))
            return nullptr;

        if (!JsonString)
            return nullptr;

        const jstring Result = JSStringToJavaString(Environment, JsonString);
        JSStringRelease(JsonString);

        return Result;
    }

    /*
     * Class:     me_ayydxn_luminescence_javascript_JSValue_NativeMethods
     * Method:    nJSValueToString
     * Signature: (JJ)Ljava/lang/String;
     */
    jstring JNICALL JSValueToString_Native(JNIEnv* Environment, jclass, jlong ContextHandle, jlong ValueHandle)
    {
        const JSStringRef StringRef =
            JSValueToStringCopy(ToContextRef(ContextHandle), ToValueRef(ValueHandle), nullptr);
        if (!StringRef)
            return Environment->NewStringUTF("");

        const jstring Result = JSStringToJavaString(Environment, StringRef);
        JSStringRelease(StringRef);

        return Result;
    }
}

static constexpr JNINativeMethod JSValueMethods[] = {
    JNI_METHOD("nJSValueGetType", "(JJ)Lme/ayydxn/luminescence/javascript/JSType;", JSValueGetType_Native),
    JNI_METHOD("nJSValueIsObject", "(JJ)Z", JSValueIsObject_Native),
    JNI_METHOD("nJSValueIsArray", "(JJ)Z", JSValueIsArray_Native),
    JNI_METHOD("nJSValueIsFunction", "(JJ)Z", JSValueIsFunction_Native),
    JNI_METHOD("nJSValueToBoolean", "(JJ)Z", JSValueToBoolean_Native),
    JNI_METHOD("nJSValueToNumber", "(JJ)D", JSValueToNumber_Native),
    JNI_METHOD("nJSValueToJSONString", "(JJI)Ljava/lang/String;", JSValueToJSONString_Native),
    JNI_METHOD("nJSValueToString", "(JJ)Ljava/lang/String;", JSValueToString_Native),
};

bool Luminescence::RegisterJSValueMethods(JNIEnv* Environment)
{
    return RegisterNativeMethods(Environment, "me/ayydxn/luminescence/javascript/JSValue$NativeMethods", JNI_METHODS_AND_COUNT(JSValueMethods));
}
